File: py_tdd/py_tdd/money.py
from dataclasses import dataclass


@dataclass
class Money:
    amount: float
    currency: str

    def times(self, multiplier: int) -> "Money":
        return Money(amount=self.amount * multiplier, currency=self.currency)

    def divide(self, divisor: int) -> "Money":
        return Money(amount=self.amount / divisor, currency=self.currency)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Money):
            return self.amount == other.amount and self.currency == other.currency
        return False

    def __str__(self) -> str:
        return f"{self.currency} {self.amount:0.2f}"

File: go/stocks/money.go
package stocks

type Money struct {
	amount   float64
	currency string
}

func (m Money) Times(multiplier int) Money {
	return Money{amount: m.amount * float64(multiplier), currency: m.currency}
}

func (m Money) Divide(divisor int) Money {
	return Money{amount: m.amount / float64(divisor), currency: m.currency}
}

func (m Money) Add(other *Money) *Money {
	if m.currency == other.currency {
		m.amount += other.amount
		sum := NewMoney(m.amount, m.currency)
		return &sum
	}
	return nil
}

func NewMoney(amount float64, currency string) Money {
	return Money{amount, currency}
}
